# Laboratorio 1 · Bitácora

**Nombre: Camilo Andres Fuquene Suesca**  
**Usuario de GitHub:camilofuquene-ui**  
**Fecha:28/08/2026**  

---

> Los enunciados están en la guía del laboratorio. Aquí solo van tus
> predicciones, tus resultados y tus explicaciones.

> **La regla que hace que esto sirva de algo:** la predicción se escribe
> *antes* de ejecutar. Si la rellenas después ya sabiendo el resultado, el
> ejercicio no mide nada y tú no aprendes nada. Nadie va a comprobarlo:
> es un trato contigo mismo.


In [5]:
from rlrs.dp import greedy_policy, q_value, value_iteration
from rlrs.envs import ACTION_NAMES, GridWorld
from rlrs.evaluation import compare, evaluate
from rlrs.policies import GreedyTabularPolicy, RandomPolicy

# Este cuaderno es una bitácora: no define algoritmos, los usa.
# Si necesitas escribir una función que valga la pena conservar,
# va en src/rlrs/, no aquí.


## Mi variante

Ejecuta `uv run python scripts/variante.py` y anota lo que te tocó.


In [6]:
RUIDO  = 0.4   # <- rellena
COSTE  = -0.1   # <- rellena
GAMMA  = 0.9

mi_env = GridWorld(noise=RUIDO, step_reward=COSTE)
mi_env.noise, mi_env.step_reward


(0.4, -0.1)

---
## Ejercicio 1 · El respaldo a mano


**Antes de ejecutar.** ¿Cuál de las cuatro acciones crees que gana en (0,3), y por qué?

_Tu predicción: En (0,3), creo que gana la acción de derecha, ya que es la que lleva directamente hacia la meta_ 

¿Cuál de las cuatro acciones queda segunda, y por qué esa y no otra?

_Tu predicción: La acción que queda de segunda, seria la de arriba, ya que rebotaria quedando en el misma casilla (0,3), las otras no ya que derecha me llevaria directo a la meta, izquierda se alejaria de la meta y abajo de igual forma me alejaria de la meta y ademas me acercaria a la penalización._ 


In [7]:
# tu código
from rlrs.dp import q_value, value_iteration
from rlrs.envs import ACTION_NAMES, GridWorld

env = GridWorld()                      # el de la guía, no tu variante
valores, politica, barridos = value_iteration(env, gamma=0.9)

for a in range(env.n_actions):
    print(f'{ACTION_NAMES[a]:<10} {q_value(env, valores, (0, 3), a, 0.9):+.6f}')


arriba     +0.800017
derecha    +0.928402
abajo      +0.554337
izquierda  +0.636940


**Explica.** ¿Coincidió? ¿Por qué gana esa y no las otras?

_Tu respuesta: Si coincidio siendo la mayor **derecha**, las otras tienen un valor menor debido a que se alejan de la meta (**abajo y izquierda**) y en el caso de **arriba** rebota y queda en la misma casilla (0,3)._ 


---
## Ejercicio 2 · Tu variante, medida


**Antes de ejecutar**. Sin resolverlo: ¿tu entorno necesitará más o menos barridos que los 35 del entorno de la guía? Apuesta y di por qué.

_Tu predicción: Debido a que el ruido de mi entorno es mayor que el de la guia, los barridos aumentarian_

In [8]:
# tu código
from rlrs.evaluation import compare
from rlrs.policies import GreedyTabularPolicy, RandomPolicy

mi_env = GridWorld(noise=RUIDO, step_reward=COSTE)     # los tuyos
valores, politica, barridos = value_iteration(mi_env, gamma=0.9)
print(f'{barridos} barridos · V(3,0) = {valores[mi_env.state_index((3, 0))]:+.4f}')
print(mi_env.render_values(valores, politica))

for r in compare(mi_env,
                 [GreedyTabularPolicy(politica, name='optima'),
                  RandomPolicy(mi_env.n_actions, seed=0)],
                 episodes=300, base_seed=0):
    print(' ', r)

61 barridos · V(3,0) = -0.4037
-0.06>  +0.17>  +0.38>  +0.71>   +1    
-0.20^    ###   +0.16^  +0.14<   -1    
-0.31^  -0.25>  -0.07^    ###   -0.58v 
-0.40^  -0.35^  -0.27^  -0.38<  -0.50< 
  optima         retorno -0.328 [-0.396, -0.260]  exito 100.0%  pasos  14.3
  aleatoria      retorno -6.008 [-6.377, -5.640]  exito 25.7%  pasos  58.0


**Anota.** Barridos, V(3,0), retorno con su intervalo, tasa de éxito y pasos medios.

_Tus números: 61 barridos, -0.4037, 
  optima         retorno -0.328 [-0.396, -0.260]  exito 100.0%  pasos  14.3
  aleatoria      retorno -6.008 [-6.377, -5.640]  exito 25.7%  pasos  58.0
  
Al final si coincidio aumentaron los barridos y esto debido al aumento del ruido con respecto al de la guia_ 


---
## Ejercicio 3 · Subir gamma


**Antes de ejecutar.** Al pasar de 0,9 a 0,99: ¿qué le pasa al número de barridos? ¿Y a la política?

_Tu predicción: El numero de barridos aumentaria y la politica de igual forma cambiaria _ 


In [9]:
# tu código
for g in (0.5, 0.9, 0.99):
    v, p, n = value_iteration(mi_env, gamma=g)
    print(f'\ngamma = {g}   {n} barridos   V(3,0) = {v[mi_env.state_index((3, 0))]:+.4f}')
    print(mi_env.render_values(v, p))


gamma = 0.5   22 barridos   V(3,0) = -0.1974
-0.16>  -0.09>  +0.08>  +0.61>   +1    
-0.19^    ###   -0.09^  -0.07<   -1    
-0.19^  -0.19>  -0.16^    ###   -0.20v 
-0.20^  -0.19>  -0.19^  -0.20<  -0.20< 

gamma = 0.9   61 barridos   V(3,0) = -0.4037
-0.06>  +0.17>  +0.38>  +0.71>   +1    
-0.20^    ###   +0.16^  +0.14<   -1    
-0.31^  -0.25>  -0.07^    ###   -0.58v 
-0.40^  -0.35^  -0.27^  -0.38<  -0.50< 

gamma = 0.99   91 barridos   V(3,0) = -0.3111
+0.14>  +0.37>  +0.54>  +0.78>   +1    
-0.03^    ###   +0.36^  +0.33<   -1    
-0.17^  -0.10>  +0.12^    ###   -0.67v 
-0.31^  -0.25^  -0.14^  -0.30<  -0.51< 


**Explica.** ¿Qué se movió mucho y qué se movió poco? ¿Por qué?

_Tu respuesta: Se movio mucho el valor de las celdas, esto ya que esta completamente relacionado con el aumento de gamma, ya que su aumento significa mas demora en converger, en cuanto a la politica no hay cambios.

Mi predicción coincidio en el aumento de barridos, pero el de la politica no coincidio lo que me da a entender que el cambio de gamma no hace un cambio directo en la politica_ 


---
## Ejercicio 4 · Quitar el ruido


**Antes de ejecutar.** Con ruido 0, ¿cambia la política óptima respecto a la tuya? ¿En qué casillas?

_Tu predicción: No cambia la politica ya que la diferencia del ruido con respecto a la mia no es muy grande entonces la direccion seguiria siendo la misma_ 


In [10]:
# tu código
for ruido in (0.0, 0.2, 0.6):
    e = GridWorld(noise=ruido)
    v, p, n = value_iteration(e, gamma=0.9)
    print(f'\nruido = {ruido}   {n} barridos')
    print(e.render_values(v, p))



ruido = 0.0   9 barridos
+0.62>  +0.73>  +0.86>  +1.00>   +1    
+0.52^    ###   +0.73^  +0.86^   -1    
+0.43^  +0.52>  +0.62^    ###   +0.27v 
+0.34^  +0.43^  +0.52^  +0.43<  +0.34< 

ruido = 0.2   35 barridos
+0.48>  +0.61>  +0.75>  +0.93>   +1    
+0.37^    ###   +0.61^  +0.59^   -1    
+0.28^  +0.36>  +0.47^    ###   +0.10v 
+0.21^  +0.27^  +0.35^  +0.26<  +0.17< 

ruido = 0.6   90 barridos
+0.02^  +0.17>  +0.33>  +0.62>   +1    
-0.08^    ###   +0.21^  +0.28<   -1    
-0.13^  -0.11>  +0.01^    ###   -0.28v 
-0.18^  -0.16>  -0.13<  -0.19<  -0.25v 


**Explica.** ¿Acertaste? Si te sorprendió, di exactamente qué esperabas y qué pasó.

_Tu respuesta: Si acerte, esperaba que no cambiara, pero con un poco mas de ruido como es el caso de 0.6, si me sorprendi al ver que la en algunas celdas la politica no tuviera el sentido mas logico _ 


---
## Ejercicio 5 · Encarecer el paso


**Antes de ejecutar.** Con coste por paso −2, ¿qué hará el agente?

_Tu predicción: yo digo que el agente tendria menos exito_ 


In [11]:
# tu código
for coste in (-0.001, -0.04, -2.0):
    e = GridWorld(step_reward=coste)
    v, p, n = value_iteration(e, gamma=0.9)
    ev = evaluate(GridWorld(step_reward=coste), GreedyTabularPolicy(p), episodes=300, base_seed=0)
    print(f'coste {coste:>7} · {n:>2} barridos · {ev}')
    print(e.render_values(v, p), '\n')

coste  -0.001 · 41 barridos · avida          retorno +0.991 [+0.991, +0.992]  exito 100.0%  pasos   9.6
+0.62>  +0.71>  +0.82>  +0.94>   +1    
+0.54^    ###   +0.71^  +0.65<   -1    
+0.48^  +0.53>  +0.61^    ###   +0.35v 
+0.42^  +0.47^  +0.52^  +0.46<  +0.40<  

coste   -0.04 · 35 barridos · avida          retorno +0.636 [+0.603, +0.670]  exito 98.0%  pasos   9.1
+0.48>  +0.61>  +0.75>  +0.93>   +1    
+0.37^    ###   +0.61^  +0.59^   -1    
+0.28^  +0.36>  +0.47^    ###   +0.10v 
+0.21^  +0.27^  +0.35^  +0.26<  +0.17<  

coste    -2.0 · 29 barridos · avida          retorno -13.393 [-13.744, -13.043]  exito  6.0%  pasos   7.3
-6.54>  -4.47>  -2.31>  +0.31>   +1    
-8.18^    ###   -3.66>  -1.29>   -1    
-9.20>  -7.71>  -5.86^    ###   -1.46^ 
-10.11>  -8.85>  -7.44>  -5.90>  -3.94^  



**Explica.** ¿Qué está optimizando exactamente el agente para comportarse así?

_Tu respuesta: si esperaba que la tasa de exito fuera mucho menor en -2.0, pero no con tan poco porcentaje, en eso si me sorprendio y supongo que maximiza el camino mas corto por que usa menos pasos, por lo que de igual forma esta cambiando la politica_ 


---
## Ejercicio 6 · El error plantado


**Antes de ejecutar**. Antes de ejecutarlo: en el intento 2, con 
γ = 1 y coste =−0,04, ¿crees que converge o no? Escríbelo.

_Tu respuesta: No creo ya que al ser gamma igual a 1 no deberia porque converger.

UPTC · Sesion 1 · Que sostiene la convergencia

  1) value_iteration(env, gamma=1.0)
     ValueError: gamma debe estar en [0, 1); se recibio 1.0
     La guardia protege una garantia: sin gamma < 1 el operador de
     Bellman deja de ser una contraccion.

  2) gamma = 1.0, recompensa por paso -0.04  (el entorno de siempre)
   barrido       V(3,0)       max|V|       cambio
  ------------------------------------------------
         1      -0.0400       0.7920     0.792000
         5      -0.2000       0.9513     0.337498
        10       0.4938       0.9685     0.158584
        50       0.6675       0.9721     0.000000
       100       0.6675       0.9721     0.000000
       500       0.6675       0.9721     0.000000
      1000       0.6675       0.9721     0.000000
      2000       0.6675       0.9721     0.000000
     Converge. Quedarse dando vueltas cuesta -0.04 por paso, o sea
     -infinito, asi que ninguna politica que el max prefiera lo hace.
     Es un camino mas corto estocastico, y ahi gamma = 1 esta bien
     definido. Perder la garantia no es perder la convergencia.

  3) gamma = 1.0, recompensa por paso +0.01  (ahora le pagamos por moverse)
   barrido       V(3,0)       max|V|       cambio
  ------------------------------------------------
         1       0.0100       0.8020     0.802000
         5       0.0500       0.9703     0.368426
        10       0.8791       1.0174     0.180948
        50       1.4173       1.4173     0.010000
       100       1.9173       1.9173     0.010000
       500       5.9173       5.9173     0.010000
      1000      10.9173      10.9173     0.010000
      2000      20.9173      20.9173     0.010000
     No converge. A partir del barrido 100 los valores crecen +0.01
     por barrido, indefinidamente, y el cambio se estanca en 0.01:
     el criterio de parada nunca se dispara.

  El mismo entorno con gamma = 0.9 converge en 42 barridos,
  con max|V| = 0.9496. La unica diferencia es el descuento.

  El diagnostico tiene dos capas.
    Matematica: con gamma = 1 existe una politica que nunca termina y
    acumula +0.01 sin fin, luego su retorno es +infinito. No hay punto
    fijo finito al que converger.
    De diseno: el error no fue poner gamma = 1, fue pagar por el
    proceso en vez de por el resultado. El descuento solo lo tapaba.

**Explica las dos capas del diagnóstico.**

_La capa matemática: No converge poque si la politica permite obtener sin limite la recompensa de +0.01, el retorno acumulado creceria hasta el infinito_ 

_La capa de diseño: No es logico recompensar por cada paso que se de ya que para intentar aumentar el retorno, el agente se quedaria en el proceso indefinidamente en lugar de tratar de llegar a la meta_ 


---
## Cierre

**Lo que más me sorprendió hoy: La forma clara de ver como cada parametro me afecta el resultado del agente, y la parte donde explica que premiar por el proceso puede generar que nunca se llegue a la meta.** 

**Lo que todavía no entiendo: No entiendo la forma en la que se calcula o se sabe la politica optima en cada casilla.** 
